src/logic/nontrivial.lean

```diff
@@ -128,3 +133,59 @@ begin
   { exact ⟨x₁, (hf.ne_iff' h).2 hx⟩ },
   { exact ⟨x₂, h⟩ }
 end
+
+namespace tactic
+
+/--
+Given a goal `a = b` or `a ≤ b` in a type `α`, generates an additional hypothesis `nontrivial α`
+(as otherwise `α` is a subsingleton and the goal is trivial).
+-/
+meta def nontriviality_by_elim : tactic unit :=
+do
+  t ← (do `(%%a = %%b) ← target >>= whnf, infer_type a) <|> (do `(%%a ≤ %%b) ← target, infer_type a) <|>
+    fail "Goal is not `_ = _` or `_ ≤ _`",
+  alternative ← to_expr ``(subsingleton_or_nontrivial %%t),
+  n ← get_unused_name "_inst",
+  tactic.cases alternative [n, n],
+  `[{ resetI, try { apply le_of_eq }, apply subsingleton.elim, }] <|>
+    fail format!"Could not prove goal assuming `subsingleton {t}`",
+  reset_instance_cache
+
+/--
+Given a goal `a ≠ b` or `a < b` in a type `α`, tries to generate a `nontrivial α`
+hypothesis from existing hypotheses using `nontrivial_of_ne` and `nontrivial_of_lt`.
+-/
+meta def nontriviality_by_assumption : tactic unit :=
+do
+  t ← (do `(%%a ≠ %%b) ← target, infer_type a) <|> (do `(%%a < %%b) ← target, infer_type a) <|>
+    fail "Goal is not `_ ≠ _` or `_ < _`",
```

Comment:
On the previous branch Gabriel had suggested deleting these two lines.

If I do this, we don't know which type to look for an instance of `nontrivial` for. It would be plausible to just look for any, by using a placeholder, but I'd worry about finding the wrong one.

Reply:
Well, this failure message will never display in `notriviality`. Not sure if that's what Gabriel was pointing out.

You can get `t` more cleanly with `` `(@ne %%t _ _) ← target``

Reply:
Is there a way to get `t` without using `infer_type` when I need to consider two possibilities `≠` and `<`? I can't get around using `<|>`. 

Is ``(do `(@ne  %%t _ _) ← target, return t)`` any better than ``(do `(%%a ≠ %%b) ← target, infer_type a)``?

Reply:
You can do something like 
```lean
guardb $ e.is_app_of `ne || e.is_app_of `has_le.le,
return $ e.get_app_args.head
```
but looking at this again, would it be better to unify with `ne` or `lt` instead of checking syntactically, to capture `¬ a = b`, `>`, etc? 



> Is `` (do `(@ne %%t _ _) ← target, return t) `` any better than `` (do `(%%a ≠ %%b) ← target, infer_type a) ``?

It's not a particularly big deal but it seems better to avoid a call to `infer_type` when the type info is already in the term.

Reply:
For the record, what I wanted to say is that the tactic shouldn't be limited to `=` and `≤` goals.  Maybe it should take an optional argument, like the `inhabit` tactic?  That is, `nontriviality α`?

Reply:
Okay, I've switched to using `unify`.

I'm not entirely sure what the behaviour should be for other goals (besides `=` and `≤`). Are we going to branch into cases and create a goal with a `subsingleton` hypothesis as well?

Reply:
There's nothing special about the goal in the ne/lt case, is there? It just uses the goal to pick a type and then tries to prove nontriviality using hypotheses. `nontriviality_by_assumption` wouldn't need to be restricted to certain goals if it knew what type to look for.

I agree with Gabriel that the optional argument makes sense. If it's provided and the goal is eq/le and the types match, ignore it. If it's provided otherwise, do `nontriviality_by_assumption` except skip the first line. Otherwise do what you do now.

Reply:
Okay, I've revised the tactic, so it now takes an optional type argument.